L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorCoarsenerImpl.cc

```diff
@@ -0,0 +1,73 @@
+#include "L1Trigger/L1THGCal/interface/concentrator/HGCalConcentratorCoarsenerImpl.h"
+
+HGCalConcentratorCoarsenerImpl::
+HGCalConcentratorCoarsenerImpl(const edm::ParameterSet& conf)
+  : fixedDataSizePerHGCROC_(conf.getParameter<bool>("fixedDataSizePerHGCROC")),
+    coarseTCmapping_(std::vector<unsigned>{HGCalCoarseTriggerCellMapping::kCTCsizeVeryFine_,HGCalCoarseTriggerCellMapping::kCTCsizeVeryFine_,HGCalCoarseTriggerCellMapping::kCTCsizeVeryFine_,HGCalCoarseTriggerCellMapping::kCTCsizeVeryFine_})
+{    
+}
+
+
+void 
+HGCalConcentratorCoarsenerImpl::
+updateCoarseTriggerCellMaps(const l1t::HGCalTriggerCell& tc, uint32_t ctcid){
+  
+  auto & ctc = coarseTCs_[ctcid];
+  
+  ctc.sumPt += tc.pt();
+  ctc.sumHwPt += tc.hwPt();
+  ctc.sumMipPt += tc.mipPt();
+
+  if ( tc.mipPt() > coarseTCs_[ctcid].maxMipPt ){
+    ctc.maxId = tc.detId();
+    ctc.maxMipPt = tc.mipPt();
+  }
+
+}
+
+void
+HGCalConcentratorCoarsenerImpl::
+assignCoarseTriggerCellEnergy(l1t::HGCalTriggerCell &tc, uint32_t ctcid){
+  auto & ctc = coarseTCs_[ctcid];
+  tc.setHwPt(ctc.sumHwPt);
+  tc.setMipPt(ctc.sumMipPt);
+  tc.setPt(ctc.sumPt);
+}
+
+void 
+HGCalConcentratorCoarsenerImpl::
+coarsen(const std::vector<l1t::HGCalTriggerCell>& trigCellVecInput, std::vector<l1t::HGCalTriggerCell>& trigCellVecOutput)
+{ 
+  coarseTCs_.clear();
+
+  // first pass, fill the coarse trigger cell information
+  for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
+    int thickness = triggerTools_.thicknessIndex(tc.detId(),true);
+    if ( triggerTools_.isSilicon(tc.detId() ) ){
+      thickness = triggerTools_.thicknessIndex(tc.detId(),true);
+    }else if ( triggerTools_.isScintillator(tc.detId()) ){
+      thickness = 3;
+    }
+    
+    if ( fixedDataSizePerHGCROC_ && thickness == kHighDensityThickness_ ){
+      trigCellVecOutput.push_back( tc );
+      continue;
+    }
+
+    uint32_t ctcid = coarseTCmapping_.getCoarseTriggerCellId(tc.detId());
+    updateCoarseTriggerCellMaps( tc, ctcid );
+  }
+
+  for (auto & ctc : coarseTCs_){
+   l1t::HGCalTriggerCell triggerCell;   
+   assignCoarseTriggerCellEnergy( triggerCell, ctc.first );
+   uint32_t evenId =  coarseTCmapping_.getEvenDetId( ctc.second.maxId );       
+   GlobalPoint point = coarseTCmapping_.getCoarseTriggerCellPosition( triggerCell.detId() );
```

Comment:
True about the initialisation.
Good point about possibly passing the ctcId - but then this would mean equivalently more computation in `HGCalConcentratorSuperTriggerCellImpl` where the function is also used.

Reply:
Can you point me to the code in `HGCalConcentratorSuperTriggerCellImpl`?

Reply:
https://github.com/snwebb/cmssw/blob/aae3872fe733046f5070e06f24ece8da8d8ddccd/L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorSuperTriggerCellImpl.cc#L62

Reply:
Actually it's the same thing, the stc is also available there and the `getCoarseTriggerCellPosition` can be called outside the `output_ids` loop.

Reply:
Not quite, since we have the `stc`, but need the `ctc` (i.e for each `output_id` we will have a different value).
I agree it's potentially confusing.

Reply:
Ok, it's the coarseTC mapping, not the superTC mapping. But there won't be extra cost in that case moving the TC->CTC outside the `getCoarseTriggerCellPosition` function.

Reply:
Ok done.